### PR TITLE
Added a generic convert function strategy

### PR DESCRIPTION
**Pull request by jesper-friis**
# Description:
Added a generic convert function strategy. 

It calls a Python function to convert zero or more input instances to zero or more output instances. 

The module containing the Python function must exists in your PYTHONPATH.

## Type of change:

- [ ] Bug fix.
- [x] New feature.
- [ ] Documentation update.

## Checklist for the reviewer:

This checklist should be used as a help for the reviewer.

- [ ] Is the change limited to one issue?
- [ ] Does this PR close the issue?
- [ ] Is the code easy to read and understand, including clearly named variables?
- [ ] Do all new feature have an accompanying new test?
- [ ] Has the documentation been updated as necessary?

### CODE REVIEW

File: .github/workflows/ci_tests.yml [hunk 1]
```diff
@@ -64,6 +64,10 @@ jobs:
         pip install -e .
         pip install safety
 
+    - name: List installed packages
+      run: |
+        pip list
```
Comment (francescalb) [inline .github/workflows/ci_tests.yml:69]: Why ddo you want this?

File: .github/workflows/ci_tests.yml [hunk 2]
```diff
@@ -100,7 +106,8 @@ jobs:
       run: |
         python -m pip install -U pip
         pip install -U setuptools wheel
-        pip install -e .[dev]
+        pip install -U -r requirements.txt -r requirements_dev.txt
+        pip install -e .
```
Comment (francescalb) [inline .github/workflows/ci_tests.yml:110]: Why this change? Doesn't .[dev] isntall all standard requirements + dev requirements?

File: .github/workflows/ci_tests.yml [hunk 3]
```diff
@@ -184,7 +194,8 @@ jobs:
       run: |
         python -m pip install -U pip
         pip install -U setuptools wheel
-        pip install -e .[docs]
+        pip install -U -r requirements_docs.txt
+        pip install -e .
```
Comment (francescalb) [inline .github/workflows/ci_tests.yml:198]: Same comment as above

File: requirements.txt [hunk 4]
```diff
@@ -1,6 +1,7 @@
-DLite-Python>=0.3.3,<1
+DLite-Python>=0.3.3,<0.4
```
Comment (francescalb) [inline requirements.txt:1]: Why not latest dlite-python?

File: tests/strategies/paths.py
Comment (francescalb) [inline tests/strategies/paths.py:1]: What is the purpose of this test? It is not really testing anything except that it is possible to add paths to previous paths.
(sys should be tested in sys and dlite should be tested in dlite)

**Review by francescalb — CHANGES_REQUESTED** [review]
I have a hard time understanding the purpose of this converter strategy, and how to use it. What are the imput instances, what are the output instances?

Can you be more explicit in the documentation? I have not properly reviewed the code yet bacuase I am confused about what is supposed to do what here.

File: .github/workflows/ci_tests.yml (hunk 1 above)
Comment (jesper-friis) [reply .github/workflows/ci_tests.yml:69]: I added it for debugging purposes to see what version of all packages that are installed on GitHub (which might differ from my local machine). 

It doesn't take any time, so I thought that it might be useful to keep to make it easier to spot dependency issues on GitHub. But we can also remove it again if you want.

File: .github/workflows/ci_tests.yml (hunk 2 above)
Comment (jesper-friis) [reply .github/workflows/ci_tests.yml:110]: I splittet it into two lines, because I want to install the requirements with the update (-U) option, which doesn't make sense for development (-e .) installation.

File: .github/workflows/ci_tests.yml (hunk 3 above)
Comment (jesper-friis) [reply .github/workflows/ci_tests.yml:198]: ...and the same answer as above ;-)

File: requirements.txt (hunk 4 above)
Comment (jesper-friis) [reply requirements.txt:1]: Because it leads to segfault in codecov so that the CI/CD won't run on GitHub.

I haven't had time to dig into exactly why it fails, but there is an old bug that leads to segfault when iterating over storages. I tried to look at it before releasing v0.4.0, but was not able to solve the issue. My guess is that the changes I made, made the bug slightly more exposed such codecov is now able to find and run the problematic code and thereby segfaults. 

It is definitely a bug that should be fixed. But iterating over storages is not something we have documented or is a critical feature for using dlite. Still, I think that it will be very useful to be able to easily list all available storages-

File: tests/strategies/paths.py
Comment (jesper-friis) [reply tests/strategies/paths.py:1]: This is not really a test, just a helper module for setting up paths correct when running the tests outside pytest.

Pytest is nice, but does a lot of magic. When debugging, it is very helpful to be able to run the tests without pytest to better know what is going on.

File: .github/workflows/ci_tests.yml [hunk 5]
```diff
@@ -119,6 +126,8 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
+        # There seems to be an issue with module search in Python 3.11
+        # python-version: ["3.9", "3.10", "3.11"]
         python-version: ["3.9", "3.10"]
```
Comment (quaat) [inline .github/workflows/ci_tests.yml:131]: Is there an issue for addressing the 3.11 problems?

File: .github/workflows/ci_tests.yml [hunk 6]
```diff
@@ -135,7 +144,8 @@ jobs:
       run: |
         python -m pip install -U pip
         pip install -U setuptools wheel
-        pip install -e .[dev]
```
Comment (quaat) [inline .github/workflows/ci_tests.yml:138]: why are 'requirements.*' preferred over setup.py?

File: oteapi_dlite/strategies/convert.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+        This method will be called through the strategy-specific endpoint
+        of the OTE-API Services.
+
+        Parameters:
+            session: A session-specific dictionary context.
+
+        Returns:
+            SessionUpdate instance.
+        """
+        config = self.convert_config.configuration
```
Comment (quaat) [inline oteapi_dlite/strategies/convert.py:131]: Remember to update the config with the relevant fields from the session

**Review by quaat — CHANGES_REQUESTED** [review]

File: .github/workflows/ci_tests.yml (hunk 5 above)
Comment (jesper-friis) [reply .github/workflows/ci_tests.yml:131]: Added issue #168

File: .github/workflows/ci_tests.yml (hunk 6 above)
Comment (jesper-friis) [reply .github/workflows/ci_tests.yml:138]: Copy of my answer to Francesca:

I splittet it into two lines, because I wanted to install the requirements with the update (-U) option, which doesn't make sense for development (-e .) installation.

But it might not be necessary. I tried to change a lot of things before the CI on GitHub finally went through. It could very well be the change in line 91-92 that did the work...

File: .github/workflows/ci_tests.yml (hunk 1 above)
Comment (francescalb) [reply .github/workflows/ci_tests.yml:69]: Your choice :)

File: oteapi_dlite/strategies/convert.py (hunk 7 above)
Comment (jesper-friis) [reply oteapi_dlite/strategies/convert.py:131]: Yes, good point. That is needed if we want a later filter to have an effect. 

I was thinking about encouraging not specifying labels in the configuration. But fetching the label from the session could be useful. However, a single `label` in the session will not do, since each input and output has an optional label, so we need to be more specific when specifying labels in the session. 

Would it make sense to allow variable substitutions in the configuration of a partial pipeline, like

```yaml
  https://www.ntnu.edu/physmet/data#image_analyser:
    function:
      functionType: application/vnd.dlite-convert
      configuration:
        module_name: temdata.image_analyser
        function_name: image_analyser
        inputs:
          - datamodel: http://onto-ns.com/meta/0.1/TEMImage
            label: ${temimage}
        outputs:
          - datamodel: http://onto-ns.com/meta/0.1/PrecipitateStatistics
            label: ${precipitate_statistics}
    mapping:
      mappingType: mappings
      prefixes:
       ...
      triples:
        ...
```

where `${temimage}` and `${precipitate_statistics}` are substituted from the session. 

On the *pros* side, this would allow templated partial pipelines with improved re-usability. 

On the *conc* side, it will be an extra layer of complexity.

Comment (quaat) [reply oteapi_dlite/strategies/convert.py:131]: I am not completely sold on the idea of labels. Also, if we start creating declarative partial pipelines with variables, they become hard to exchange/share between instances. Why not refer to a resource directly?

**Review by quaat — APPROVED** [review]
I am not convinced we need labels or variables in the declarative pipelines, but this might be a discussion for the strategy meeting. Otherwise, merge as you wish

File: oteapi_dlite/strategies/convert.py (hunk 7 above)
Comment (jesper-friis) [reply oteapi_dlite/strategies/convert.py:131]: A agree that a partial pipeline should document a single data source and in general not contain variables. However, there are cases where fetching parameters from the configuration are useful. One case is a partial pipeline documenting a SQL database. The documentation of the database should be fix, while the query may vary each time we execute the pipeline. Another possible use case is to specify the label a parser should use when storing an newly created instance into the collection and correspondingly, the label a generator should use when fetching an instance from the collection. In this case the labels may be variables when documenting the partial pipelines, but must be assigned and internally consistent before executing the full pipeline. 
Note: often we don't need
